### ISSUE TITLE
stack overflow on Template Literal Types

### ISSUE THREAD

**Issue by TechQuery** [TypeStrong/typedoc#1440]
## Expected Behavior

generate document for [Template Literal Types][1] like this:

```TypeScript
interface CustomElement {
    tagName: `${string}-${string}`
}
```

## Actual Behavior

> TypeDoc exiting with unexpected error:
> RangeError: Maximum call stack size exceeded
>     at getBaseConstraintOfType (node_modules/typescript/lib/typescript.js:53240:41)      
>     at getTypeFacts (node_modules/typescript/lib/typescript.js:62711:37)
>     at getTypeFacts (node_modules/typescript/lib/typescript.js:62711:24)
>     at getTypeFacts (node_modules/typescript/lib/typescript.js:62711:24)

## `tsconfig.json`

```JSON
{
    "compilerOptions": {
        "outDir": "dist/",
        "target": "ES5",
        "sourceMap": true,
        "declaration": true,
        "allowJs": true,
        "noImplicitReturns": true,
        "downlevelIteration": true,
        "module": "ES2020",
        "moduleResolution": "Node",
        "esModuleInterop": true,
        "experimentalDecorators": true,
        "jsx": "react",
        "jsxFactory": "createCell",
        "jsxFragmentFactory": "Fragment",
        "lib": ["ES2020", "DOM"]
    },
    "include": ["./source/**/*"],
    "typedocOptions": {
        "name": "WebCell",
        "excludeExternals": true,
        "excludePrivate": true,
        "readme": "./ReadMe.md",
        "out": "docs/"
    }
}
```

## Environment

-   Typedoc version: 0.20.7
-   TypeScript version: 4.1.3
-   Node.js version: 14.15.0
-   OS: Windows 10 Pro 2004

[1]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-1.html#template-literal-types

**Comment by Gerrit0** [comment]
Looks like this is a problem with TypeScript, as you surmised. This *might* work in 0.20.5 since that version uses the type node instead of the type for properties. (Which causes a bug when dealing with genetics)

**Comment by Gerrit0** [comment]
I did, however, just create a new project to double check... and this seems to be working properly.

```bash
yarn init -y
yarn add typescript typedoc
vim tsconfig.json # copied in from OP
mkdir source
vim source/index.ts # copied in from OP
npx typedoc source/index.ts
```

This ran, and generated a page containing:

![image](https://user-images.githubusercontent.com/19329837/103443840-061a8d80-4c20-11eb-9aa2-785369c838c0.png)

**Comment by TechQuery** [comment]
I push a new branch of my project for your checking: https://github.com/EasyWebApp/WebCell/tree/type-optimize

`0.20.7` did throw the error as CI logs shown: https://github.com/EasyWebApp/WebCell/runs/1634386325#step:4:57

**Comment by Gerrit0** [comment]
Thanks! I pulled it down and can confirm it's broken in TypeScript, nothing I can do to fix it here... 

Good catch with the exit code being `0`!  Fixed in 0.20.8

**Comment by Gerrit0** [comment]
Looks like this will be fixed in 4.2, closing as there's nothing to be done here to fix besides wait for 4.2 to release

https://github.com/microsoft/TypeScript/issues/41651#issuecomment-758630192
